backend/src/load.rs

```diff
@@ -0,0 +1,418 @@
+                if let Some(index) = index {
+                    c_benchmarks_add += 1;
+                    let crate_name = times[0]["crate"].as_str().unwrap();
+                    data_benchmarks[index].by_crate.insert(test_name.to_string(),
+                        make_times(times, false).remove(crate_name).unwrap());
+                } else {
+                    c_benchmarks += 1;
+                    data_benchmarks.push(Data::new(date, &header, times, false));
+                }
+            }
+
+            for timing in times.members() {
+                for (phase, _) in timing["times"].entries() {
+                    phase_list.insert(phase.to_string());
+                }
+            }
+
+            if last_date.is_none() || last_date.as_ref().unwrap() < &date {
+                last_date = Some(date);
+            }
+        }
+
+        let last_date = last_date.expect("No dates found");
+
+        data_rustc.sort();
+        data_benchmarks.sort();
+
+        // Post processing to generate the summary data.
+        let summary_rustc = Summary::new(&data_rustc, last_date);
+        let summary_benchmarks = Summary::new(&data_benchmarks, last_date);
+
+        println!("read {} files", file_count);
+        println!("found {} files without times", no_times);
+        println!("{} bootstrap times", c_bootstraps);
+        println!("{} benchmarks times", c_benchmarks);
+        println!("{} benchmarks times (appended)", c_benchmarks_add);
+
+        Ok(ServerData {
+            summary_rustc: summary_rustc,
+            summary_benchmarks: summary_benchmarks,
+            crate_list: crate_list,
+            phase_list: phase_list,
+            benchmarks: benchmarks,
+            last_date: last_date,
+            data_rustc: data_rustc,
+            data_benchmarks: data_benchmarks,
+        })
+    }
+
+    pub fn by_kind(&self, kind: &str) -> &[Data] {
+        if kind == "rustc" {
+            &self.data_rustc
+        } else if kind == "benchmarks" {
+            &self.data_benchmarks
+        } else {
+            panic!("Unknown kind: {}", kind);
+        }
+    }
+}
+
+/// The data loaded for a single date, and all associated crates.
+pub struct Data {
+    pub date: NaiveDateTime,
+    pub commit: String,
+
+    /// Map of crate names to a map of phases to timings per phase.
+    pub by_crate: HashMap<String, HashMap<String, Timing>>
+}
+
+impl PartialEq for Data {
+    fn eq(&self, other: &Data) -> bool {
+        self.commit == other.commit && self.date == other.date
+    }
+}
+
+impl Eq for Data {}
+
+impl PartialOrd for Data {
+    fn partial_cmp(&self, other: &Data) -> Option<Ordering> {
+        Some(<Data as Ord>::cmp(self, other))
+    }
+}
+
+impl Ord for Data {
+    fn cmp(&self, other: &Data) -> Ordering {
+        self.date.cmp(&other.date)
+    }
+}
+
+impl Data {
+    // TODO: Document what rustc_crate means.
```

Comment:
I think that if it is true, it means the data is from a run on the bootstrap (i.e., the compiler compiling itself), rather than a benchmark